src/main/java/org/folio/rest/service/ActionService.java

```diff
@@ -0,0 +1,178 @@
+  }
+
+  private Future<ActionContext> performAction(Action action, String accountId,
+    ActionRequest request) {
+
+    return succeededFuture(new ActionContext(accountId, request, action))
+      .compose(this::findAccount)
+      .compose(this::validateAction)
+      .compose(this::createFeeFineAction)
+      .compose(this::updateAccount)
+      .compose(this::sendPatronNotice);
+  }
+
+  private Future<ActionContext> findAccount(ActionContext context) {
+    return accountRepository.getAccountById(context.getAccountId())
+      .map(context::withAccount);
+  }
+
+  private Future<ActionContext> validateAction(ActionContext context) {
+    final String amount = context.getRequest().getAmount();
+
+    return validationService.validate(context.getAccount(), amount)
+      .map(result -> context.withRequestedAmount(parseDouble(amount)));
+  }
+
+  private Future<ActionContext> createFeeFineAction(ActionContext context) {
+    ActionRequest request = context.getRequest();
+    Account account = context.getAccount();
+    Action action = context.getAction();
+    Double requestedAmount = context.getRequestedAmount();
+    double remainingAmount = account.getRemaining() - requestedAmount;
+
+    String actionType = remainingAmount == 0
+      ? action.getFullResult()
+      : action.getPartialResult();
+
+    Feefineaction feeFineAction = new Feefineaction()
+      .withAmountAction(requestedAmount)
+      .withComments(request.getComments())
+      .withNotify(request.getNotifyPatron())
+      .withTransactionInformation(request.getTransactionInfo())
+      .withCreatedAt(request.getServicePointId())
+      .withSource(request.getUserName())
+      .withPaymentMethod(request.getPaymentMethod())
+      .withAccountId(context.getAccountId())
+      .withUserId(account.getUserId())
+      .withBalance(remainingAmount)
+      .withTypeAction(actionType)
+      .withId(UUID.randomUUID().toString())
+      .withDateAction(new Date())
+      .withAccountId(context.getAccountId());
+
+    return actionRepository.save(feeFineAction)
+      .map(context.withFeeFineAction(feeFineAction));
+  }
+
+  private Future<ActionContext> updateAccount(ActionContext context) {
+    final Feefineaction feeFineAction = context.getFeeFineAction();
+    final Account account = context.getAccount();
+
+    if (isTerminalStatus(feeFineAction.getTypeAction())) {
+      account.getStatus().setName(FeeFineStatus.CLOSED.getValue());
+    }
+
+    account.setRemaining(feeFineAction.getBalance());
+    account.getPaymentStatus().setName(feeFineAction.getTypeAction());
+
+    return accountUpdateService.updateAccount(account)
+      .map(context);
+  }
+
+  private Future<ActionContext> sendPatronNotice(ActionContext context) {
+    if (isTrue(context.getRequest().getNotifyPatron())) {
+      patronNoticeService.sendPatronNotice(context.getFeeFineAction());
+    }
+    return succeededFuture(context);
+  }
+
+  public static class ActionContext {
```

Comment:
Do you use it from another class?

Reply:
Yes, `AccountsAPI`.

Reply:
I can't find using ActionContext in AccountsAPI

Reply:
https://github.com/folio-org/mod-feesfines/blob/9c491a976bac5636c85b55fdef3e655c471b448d/src/main/java/org/folio/rest/impl/AccountsAPI.java#L403